superset/alfred/utils.py

```diff
@@ -0,0 +1,325 @@
+
+    # Return the URL
+    retention_url = (
+        InstanceUtil.build_url(alfred_instance)
+        .replace(":9488", "")
+        .replace("/rest", f"/ui/{retention.retention.uri}")
+    )
+
+    return retention_url
+
+
+def sanitize_results(data):
+    result = dict()
+    if isinstance(data, list):
+        json_array = []
+        for v in data:
+            json_array.append(sanitize_results(v))
+        return json_array
+    elif isinstance(data, NamedRowTuple):
+        for key, value in zip(data._names, data):
+            result[key] = sanitize_results(value)
+    else:
+        return data
+    return result
+
+def retain_eml_to_alfred(ids, alfred_env, access_token, dates=None):
+    try:
+        client = VaultClient()
+
+        ids_string = ""
+        datetimes = []
+        if ids:
+            logger.info(f'Found eml_ids: "{ids_string}" in arguments')
+            ids_string = "','".join(ids)
+        else:
+            raise Exception("No field 'email_ids' found in request")
+        if dates:
+            logger.info(f'Found dates: "{dates}" in arguments')
+            for date in dates:
+                try:
+                    datetimes.append(datetime.strptime(date, "%Y-%m-%d"))
+                except ValueError as ve:
+                    logger.info(ve)
+                    continue
+        if alfred_env:
+            logger.info(f'Found Alfred-env: "{alfred_env}" in arguments')
+        else:
+            raise Exception("No field 'alfred_env' found in request")
+
+        logger.info("Generating access token for trino...")
+        trino_app = client.get_oauth_client("trino")
+
+        trino_acces_token, trino_ctx = client.on_behalf_of(
+            f"{trino_app.audience}/{trino_app.scope}",
+            access_token,
+            token_client_name="superset",
+        )
+
+        trino_host = os.environ.get('TRINO_HOST')
+        if not trino_host:
+            logger.error("TRINO_HOST environment variable not set")
+            raise Exception('TRINO_HOST environment variable not set')
+        logger.info(f"Establishing connection to trino at {trino_host}...")
+        conn = connect(
+            auth=JWTAuthentication(trino_acces_token),
+            http_scheme="https",
+            host=trino_host,
+            port=443,
+            catalog="hogwarts_pb",
+        )
+
+        sql = f"""
+    SELECT id, urls, "from", to, message_id, reply_to, stream_id, subject, eml_path, bcc, cc, ministerial_authorization, classification, time
+    FROM  hogwarts_pb.harmonized.eml_metadata
+    WHERE id IN ('{ids_string}')
+    """
+        if datetimes:
+            sql += " AND "
+            for i in range(len(datetimes)):
+                if i > 0:
+                    sql += " OR "
+                sql += f"(time > TIMESTAMP '{datetimes[i].strftime('%Y-%m-%d')}' AND time < TIMESTAMP '{(datetimes[i] + timedelta(days=1)).strftime('%Y-%m-%d')}')"
```

Comment:
An extra pair of brackets is required here to isolate all the date clauses together, otherwise you may end up with unexpected results if dates are different: IDs in (... , ...) AND (1st date where clause) OR (2nd date where clause) OR (3rd date where clause)